hybridq/noise/channel/channel.py

```diff
@@ -176,31 +177,120 @@ def MatrixChannel(LMatrices: tuple[array, ...],
+    # Convert norm_atol to float
+    norm_atol = float(norm_atol)
+
+    # Define sample
+    def __sample__(self, size: int = None, replace: bool = True):
+        from hybridq.utils import isintegral
+
+        # It is assumed here that s.ndim == 1
+        assert (self.s.ndim == 1)
+
+        # Get number of elements
+        idxs = np.random.choice(range(len(self.s)),
+                                size=size,
+                                replace=replace,
+                                p=self.s)
+
+        # Get gates (it is assumed here that left and right gates are the same)
+        return self.Kraus.gates[0][idxs] if isintegral(idxs) else tuple(
+            self.Kraus.gates[0][x] for x in idxs)
+
+    # Define apply
+    def __apply__(self, psi, order):
+        from hybridq.utils import dot
+
+        # It is assumed here that s.ndim == 1
+        assert (self.s.ndim == 1)
+
+        # Define how to get projection
+        def _get_projection(idx):
+            # Get s
+            s = self.s[idx]
+
+            # Get Kraus operator
+            gate = self.Kraus.gates[0][idx]
+
+            # Get projection
+            proj = dot(a=gate.matrix(),
+                       b=psi,
+                       axes=tuple(map(order.index, gate.qubits)),
+                       b_as_complex_array=not np.iscomplexobj(psi))
+
+            # Get normalization
+            norm = np.linalg.norm(proj.ravel())
+
+            # If norm is smaller than 'norm_atol', state vector is assumed to
+            # be zero
+            if norm < norm_atol:
+                norm = 0
+
+            # Get probability
+            prob = s * norm**2
+
+            # Return projection, norm and probability
+            return proj, norm, prob
+
+        # Convert order to tuple
+        order = tuple(order)
+
+        # Get random number
+        r = np.random.random()
+
+        # Initialize cumulative
+        c = 0
+
+        # Last non-zero state
+        proj, norm = None, None
+
+        # For each Kraus operator ...
+        for idx in self.__LMatrices_order:
+            # Get projection, norm and probability
+            _proj, _norm, _prob = _get_projection(idx)
+
+            # If norm is different from zero ...
+            if _norm > 0:
+                # Store state
+                proj, norm = _proj, _norm
+
+                # Update cumulative
+                c += _prob
+
+                # If the random number if smaller than the cumulative, break
+                if c >= r:
+                    break
+
+        # Check that proj and norm have been succesfully assigned
+        if proj is None or norm is None:
+            raise RuntimeError("All state vectors have a norm smaller "
+                               "than the absolute tollerance, "
+                               f"'norm_atol={norm_atol}'")
+
+        # If c is not close to one within 'norm_atol', warn the user
+        if not np.isclose(c, 1, atol=norm_atol):
```

Comment:
I dont think this is the intended check. If r=0.5 for example, c might break from the loop at the value c=0.6. This warning will almost always be raised....

I think given the previous if statement, which if it passes, guarantees we have a state with non-zero norm, we do not need to have this second check at all.